### DISCUSSION TITLE
django.db.utils.IntegrityError

### DISCUSSION THREAD

**Discussion by remchuk** [netbox-community/netbox#12817] [Help Wanted!]
Hey ! so I have been trying to bulk import  modules using csv and running into this error :
 **Key (device_id, name) = (1111, B1) already exists.**
 **duplicate key value violates unique constraint "dcim_interface_device_id_name_bffc43c4_uniq"**

Now when trying to create it manually it does work but trying to import it , it fails.

I first bulk imported it and then deleted it all , when trying to do it the second time this error happens.

Also when I created a module it received an id lets say **14** and then delete it and try again it will receive a device_id of **15** even thought no device has the id 14 already, why is that?

**Comment by candlerb** [comment]
* Can you show the CSV you're trying to import - header line plus the data line which is triggering the error?
* Can you confirm you are uploading this at `/dcim/modules/import/` ?
* What version of Netbox are you running?

Note that when you import a module, it will implicitly create copies of all the components (interfaces etc) that are defined on the module type.  I *think* that's what's happening here: you have a device that already has an interface "B1", and then you are importing a module which tries to create another interface called "B1" on the same device.

Adding column "replicate_components", "false" or "adopt_components", "true" should fix that.  (I believe the first one means "don't create any components from the module type templates", and the second means "if there's already a component with the same type and name, keep it and associate it with the module")

> Also when I created a module it received an id lets say 14 and then delete it and try again it will receive a device_id of 15 even thought no device has the id 14 already, why is that?

The module ID and the device ID are two completely different things.  Each time you create, delete and recreate a module, it will get a different (module) id.

Adding a module will never create a new device, so the device ID won't change. But it could create new components (interfaces, frontports, powerports etc) and they will have their own database IDs too.

**Reply by remchuk** [reply]
Well I will try your suggestion but the device which triggers the error has no interfaces such as we try to add for sure.
More than that is when we did it the first time it worked but after deleting it and trying again it failed. 

Meaning there is probably a bug or something else causing it .

About showing the CSV I can't but it is correct due to working the first time we tried.

We use the default headers.


NetBox version is 3.3.4

**Reply by remchuk** [reply]
> Well I will try your suggestion but the device which triggers the error has no interfaces such as we try to add for sure.
> More than that is when we did it the first time it worked but after deleting it and trying again it failed. 
> 
> Meaning there is probably a bug or something else causing it .
> 
> About showing the CSV I can't but it is correct due to working the first time we tried.
> 
> We use the default headers.
> 
> 
> NetBox version is 3.3.4

It didn't work , there is no such column as "replicate_components" you can do it manually but not with import

**Reply by candlerb** [reply]
If you won't show the CSV, then I won't attempt to help further. Sorry, but I can't work blind.

**Reply by remchuk** [reply]
It is a standard csv file having 3 columns 

Device, module bay and module type

I wish I could show you more but I'm not allowed to.

But lets say it is like that

device    | module_bay      | module_type 
server_a | module_bay_a | module_type_a

The triggered error is the device_id of server_a.

**Reply by candlerb** [reply]
FYI, in Netbox 3.5.3:

![image](https://github.com/netbox-community/netbox/assets/44789/410fd3a9-c524-4433-b61c-6efda72ac02a)

As far as I can tell, "replicate_components" and "adopt_components" were added in b3693099 in Dec 2022, which made it into [v3.3.10](https://github.com/netbox-community/netbox/releases/v3.3.10).  See feature #9361.

Therefore I suggest you first upgrade from 3.3.4 to 3.3.10. Then if you still have a problem, show how to reproduce it with fake data.

**Reply by remchuk** [reply]
It's quite a tough job to upgrade our netbox (docker-netbox) so we wait for the release of 3.5 to the docker netbox , but nevertheless the error is why does the NetBox sees a duplicate.

A solution is good but I want to find the cause and not go around it .

This bug shouldn't be happening due to the fact that all the modules were deleted yet a duplicate value is still happening.

I get that if it worked the first time means that the problem is that the modules are still stored somewhere, but it is not in the netbox nor the api.

**Reply by candlerb** [reply]
So let me paraphrase:

_"We are running an old version of Netbox, and we have a problem when uploading data to it. We are not prepared to upgrade Netbox, nor are we prepared to share any of the data which causes the problem, nor are we prepared to take the time to reproduce the problem with fake data. However, we still expect you to magically diagnose and fix our problem."_

Sorry, but that's not how [open source problem solving](http://www.catb.org/~esr/faqs/smart-questions.html#intro) works, so I'm out. `>Click<`

**Reply by remchuk** [reply]
Thats really nothing like that , I really don't understand you attitude towards that but okay.

What data can I give you that I haven't?

I will gladly reproduce it with fake data , if I would understand what it means , I'm just a starter of the Developing and using NetBox.

If you could explain what exactly you want me to send or say or do just say.

You want me to send you the CSV we try to use but change the names to something fake ? Sure thing I just don't get why.

I have no problem updating it , but it doesn't seems like a solution rather a walk around the problem don't you agree? I mean something has caused this , it worked the first time but not the second with the same data. Meaning that the data is probably saved somewhere which is not visible to the eye and I really don't know where and why

And I'm certainly doesn't expect anything, I'm just trying to find the cause , I don't expect a solution, anything helps and I'm grateful for anything to find the causing error to prevent it from happening again.

I'm sorry to cause you any trouble or take of your time . Have a good day and thanks for the help.
